bpo-41001: Add os.eventfd()

tiran opened the pull request:
Signed-off-by: Christian Heimes <christian@python.org>

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-41001](https://bugs.python.org/issue41001) -->
https://bugs.python.org/issue41001
<!-- /issue-number -->

giampaolo commented on Doc/library/os.rst:
```diff
@@ -3195,6 +3195,27 @@ features:
    .. versionadded:: 3.8
 
 
+.. function:: eventfd(initval[, flags=os.EFD_CLOEXEC])
```
What's the rationale for `flags` defaulting to `EFD_CLOEXEC` instead of `0`?
Note: I'm not familiar with this syscall, but if flags=0 is a legitimate use case then I think it makes sense to have:
```
eventfd(initval[, flags=0])
```

tiran replied:
See https://www.python.org/dev/peps/pep-0446/

CPython tries hard to make all file descriptors as ``CLOEXEC`` by default. Users can opt-out from the default with a different flag.

giampaolo replied:
Makes sense.

aeros commented on Lib/test/test_os.py:
```diff
@@ -3524,6 +3525,48 @@ def test_memfd_create(self):
 
 
+@unittest.skipUnless(hasattr(os, 'eventfd'), 'requires os.eventfd')
+@support.requires_linux_version(2, 30)
```
I think the test class should be ran with Linux 2.27+ (to at least cover `test_eventfd_initval`) and then `test_eventfd_semaphore` (or any future tests that require use of the `os.EFD_NONBLOCK` flag) can require 2.30+. I'm not sure that it's entirely necessary or the number of devices it would actually be applicable to, but it seems like a better practice to have *some* amount of test coverage for 2.27 - 2.29 since the feature will be included in those versions.

Also, would it be worthwhile to add a test that makes use of select (maybe for poll and epoll as well, via mixin), which uses them to check if the file descriptor is ready for reading or writing? I suspect those will be typically used with `os.eventfd()`, so ensuring that behavior works as expected would make sense to me. I'm just not certain if it makes sense to include those in `test_os`, or if there might be a better location.

tiran replied:
It's actually 2.6.30, not 2.30. I'll fix that with the next commit. Linux Kernel 2.6.30 was released in 2009. It's highly unlikely that somebody is running a Kernel version between 2.6.27 and 2.6.30.

aeros replied:
> It's actually 2.6.30, not 2.30. I'll fix that with the next commit.

Ah, good catch. I completely didn't notice the missing `.6`, despite recently reading over the man page for `eventfd()`. :-)

> It's highly unlikely that somebody is running a Kernel version between 2.6.27 and 2.6.30.

Yeah, after looking over the dates for the releases, I'm now in agreement that it would be highly unlikely. If anything, I suspect most on the 2.6.x branch would be using 2.6.39 or 2.6.32.

(Based on https://upload.wikimedia.org/wikipedia/en/timeline/40658da696ee210d2be5f221ac12e43b.png.)

aeros commented on Doc/library/os.rst:
```diff
@@ -3195,6 +3195,27 @@ features:
 
+.. function:: eventfd(initval[, flags=os.EFD_CLOEXEC])
+
+   Create an event file descriptor
```
I think this could be expanded upon a bit, specifically by:

1) Explicitly mentioning that the file descriptor is returned.  I think it helps to make usage a bit more clear, and is done for the above `memfd_create()`.
2) Briefly summarize the use for it as a lightweight notifier/waiter for events. Compared to a pipe, it is more resource efficient as an event notifier.

E.g.

```suggestion
   Create and return an "eventfd" file descriptor. It can be used as a lightweight
   event notifier.
   
   .. note::
      Compared to a pipe that is exclusively used for event notification,
      `eventfd()` can be used as a more resource efficient alternative that
      only uses a single file descriptor and has far less overhead.
```

While I agree with making the low-level OS functions that are thin wrappers to existing OS APIs fairly minimal in terms of documentation details (especially if they're already well documented), I think the existing one could provide a bit more information to the user that explains *why* they might want to use it. 

For the full details one can view the man page, but IMO, we at least should summarize the general purpose. Otherwise, I suspect it will significantly deter discoverability of the feature.

tiran committed:
Add eventfd_read/write, fix value type

The ``eventfd_read`` and ``eventfd_write`` functions are convenient
wrappers that perform proper read/write and conversion from/to uint64_t.

Signed-off-by: Christian Heimes <christian@python.org>

tiran committed:
Add eventfd select test case

Signed-off-by: Christian Heimes <christian@python.org>

pitrou commented on Doc/library/os.rst:
```diff
@@ -3195,6 +3195,73 @@ features:
+   :func:`eventfd_write` increments the event counter.
+
+   *initval* is the initial value of the event counter. It must be an
+   unsigned integer between 0 and 2\ :sup:`64`\ -\ 2.
```
According to this [man page](https://man7.org/linux/man-pages/man2/eventfd.2.html), the initialization value is an `unsigned int`. In any case, I don't think it's useful to document the maximum value.

tiran replied:
I have updated the documentation. I think it's useful to document that the initial value is limited to ``UINT32_MAX`` while the maximum counter is ``UINT64_MAX - 1``.

pitrou commented on Doc/library/os.rst:
```diff
@@ -3195,6 +3195,73 @@ features:
+   Provide semaphore-like semantics for reads from a :func:`eventfd` file
+   descriptor. On read the internal counter is decremented by one.
+
+   .. availability:: Linux 2.6.30 or newer
```
What's the rationale for adding an availability markup for this data member but not the others?

tiran replied:
All other items have same availability as ``eventfd``, semaphore came a bit later.

pitrou commented on Doc/library/os.rst:
```diff
@@ -3195,6 +3195,73 @@ features:
+.. function:: eventfd_read(fd)
+
+   Read value from an :func:`eventfd` file descriptor and return
+   an unsigned integer between 0 and 2\ :sup:`64`\ -\ 2. The function does
```
The maximum value needn't be documented, IMHO.

tiran replied:
IMHO it's useful. The maximum value is one smaller than UINT64_MAX.

pitrou commented on Lib/test/test_os.py:
```diff
@@ -3524,6 +3530,92 @@ def test_memfd_create(self):
+        size = 8  # read/write 8 bytes
+        initval = 42
+        fd = os.eventfd(initval)
+        self.assertNotEqual(fd, -1)
```
`self.assertGreaterEqual(fd, 0)`?

tiran replied:
Other tests use the same check. I prefer to stay consistent. -1 has a special meaning, too.

pitrou commented on Modules/posixmodule.c:
```diff
@@ -12446,6 +12451,86 @@ os_memfd_create_impl(PyObject *module, PyObject *name, unsigned int flags)
+/*[clinic end generated code: output=bebd9040bbf987f5 input=7e50f88666ba6ecc]*/
+{
+    int result;
+    if (value > (PY_ULLONG_MAX - 1)) {
```
Wouldn't you want to simply let `eventfd_write` raise the error?

tiran replied:
I have removed this check. Note that ``eventfd_write`` does not check the value. The syscall will fail with errno 22.

pitrou reviewed (changes requested):
This is nice, though I suspect `eventfd_read` does a system call even when uncontended?

tiran committed:
Adress more code reviews

- initval is a uint32_t
- improve documentation
- add whatsnew

Signed-off-by: Christian Heimes <christian@python.org>

aeros commented on Doc/whatsnew/3.10.rst:
```diff
@@ -110,6 +110,13 @@ Added the *root_dir* and *dir_fd* parameters in :func:`~glob.glob` and
+--
+
+A new :func:`os.eventfd` function and related helpers were added to wrap the
+``eventfd2`` syscall on Linux
```
Minor whatsnew suggested edit: 

```suggestion
Added a new function :func:`os.eventfd` and related helpers to wrap the ``eventfd2`` syscall
on Linux.
```

aeros reviewed (approved):
LGTM (other than a minor suggested change for the whatsnew entry). 

As mentioned in a previous review comment, I would prefer to have some mention in the docs of the specific use case for `eventfd` as a lightweight event notifier/waiter (essentially a brief summary of the "Notes" section from the man page) so that users are aware of its general functionality, even if they may not be already familiar with the system call. Unlike in the earlier years of Python, it's becoming increasingly common for developers to have started with Python before C (which has been my own experience). So, I think it's often quite helpful to at least have a summary of the purpose of the function, instead of assuming the reader is already familiar with it.

But IMO, it's fine to merge as is, that can potentially be considered as a future enhancement to the documentation. The core implementation of the wrapper, helpers, and the tests looks good to me. 

(Note that I'm not particularly familiar with the details of implementing a CPython wrapper for a system call, I mainly used the existing members that work with FDs for comparison.)

giampaolo commented on Lib/test/test_os.py:
```diff
@@ -56,6 +57,11 @@
+try:
+    import select
+except ImportError:
+    select = None
```
It seems all other modules / tests don't guard against `select` not being available. It looks like all platforms should support it (at least `select.select`).

tiran committed:
Update Doc/whatsnew/3.10.rst

Co-authored-by: Kyle Stanley <aeros167@gmail.com>

tiran committed:
Always import select

Signed-off-by: Christian Heimes <christian@python.org>

pablogsal commented on Lib/test/test_os.py:
```diff
@@ -3524,6 +3527,89 @@ def test_memfd_create(self):
+            """Pack as native uint64_t
+            """
+            return struct.pack("@Q", value)
+        size = 8  # read/write 8 bytes
```
I think this should be `struct.calcsize("L")`

tiran replied:
According to the example in the man page, it's an uint64_t in native format, so ``@Q`` is correct.

pablogsal commented on Modules/posixmodule.c:
```diff
@@ -12446,6 +12451,78 @@ os_memfd_create_impl(PyObject *module, PyObject *name, unsigned int flags)
+os.eventfd
+
+    initval: unsigned_int
+    flags: unsigned_int(bitwise=True, c_default="EFD_CLOEXEC") = EFD_CLOEXEC
```
The flags attribute is an integer in the actual sytstem call:

```
       int eventfd(unsigned int initval, int flags);
```

tiran committed:
Add example and mention man page

tiran committed:
bpo-41001: Add os.eventfd()
